### PR TITLE
Registration

### PR DESCRIPTION
**Pull request by nghamilton**
Initial registration request and response types.

### CODE REVIEW

File: consumer-data-au-api-types/consumer-data-au-api-types.cabal [hunk 1]
```diff
@@ -7,7 +7,7 @@ synopsis:            Api Types for the Australian Consumer Data Rights Specifica
 description:         See https://github.com/ConsumerDataStandardsAustralia/standards for more details.
 license:             MIT
 license-file:        LICENSE
-author:              Queensland Functional Programming Lab, Data 61, CSIRO
+author:              Queensland Functional Programming Lab, Data61, CSIRO
```
Comment (ajmcmiddlin) [inline consumer-data-au-api-types/consumer-data-au-api-types.cabal]: This is pedantic, but ideally this and the little script/shell.nix changes would be in a separate PR and get merged quickly rather than being tied to a bigger PR.

File: consumer-data-au-api-types/src/Web/ConsumerData/Au/Api/Types/Auth/Registration.hs [hunk 2]
```diff
@@ -142,6 +264,8 @@ instance FromJSON JwksUri where
         either (fail . show) pure
 
 --TODO use keyset instead of Text for JwksVal
+--TODO must include @use@ to specify either enc or sig
+--TODO must include kid
 -- | Client's JSON Web Key Set
 data JwkSet = JwksRef JwksUri --  ^ @jwks_uri@ parameter that specifies a URL for the client's JSON Web Key Set for pass-by-reference
```
[the remark this answers is not in the document]
Comment (nghamilton) [reply consumer-data-au-api-types/src/Web/ConsumerData/Au/Api/Types/Auth/Registration.hs]: I don't understand the question, can you clarify? It is a type for a JWK Set, which can be either a URI to a JWK Set, or a JWK Set.

File: consumer-data-au-api-types/src/Web/ConsumerData/Au/Api/Types/Auth/Registration.hs [hunk 3]
```diff
@@ -8,67 +8,190 @@
...
+
+isX5t :: Lens' X509ThumbPrint Bool
+isX5t f o@(X5T a)    = fmap (bool (X5T256 a) o) (f True)
+isX5t f o@(X5T256 a) = fmap (bool o (X5T a)) (f False)
+
+x509ByteString :: Lens' X509ThumbPrint ByteString
+x509ByteString f (X5T a)    = fmap X5T (f a)
+x509ByteString f (X5T256 a) = fmap X5T256 (f a)
+
+--TODO: remove all to/fromJSONs
```
[the remark this answers is not in the document]
Comment (nghamilton) [reply consumer-data-au-api-types/src/Web/ConsumerData/Au/Api/Types/Auth/Registration.hs]: This whole feature branch is still a WIP and there are a few things that aren't finished. Perhaps we should have a chat and clarify expectations of what we want to see in our PRs?

File: consumer-data-au-api-types/tests/Web/ConsumerData/Au/Api/Types/Auth/Gens.hs [hunk 4]
```diff
@@ -43,3 +56,51 @@ genClaims ::
...
+genKeyMaterial ::
+  MonadGen n
+  => n JWK.KeyMaterialGenParam
+genKeyMaterial =
+  Gen.choice
+    [ pure (JWK.ECGenParam JWK.P_256)
+    , JWK.RSAGenParam <$> Gen.int (Range.linear (2048 `div` 8) (4096 `div` 8))
+    ]
+
+sampleT :: (MonadIO m) => Gen.GenT m a  -> m a
```
Comment (ajmcmiddlin) [inline consumer-data-au-api-types/tests/Web/ConsumerData/Au/Api/Types/Auth/Gens.hs]: This isn't specific to auth, so could maybe go somewhere higher up the tree. Maybe even `Hedgehog.Extras` or something like that.

Comment (nghamilton) [reply consumer-data-au-api-types/tests/Web/ConsumerData/Au/Api/Types/Auth/Gens.hs]: Moved to Hedgehog.Helpers.
